File: python/examples/bouncing-ball-3var.py
#!/usr/bin/python
#
# Bouncing ball example
#
# Written by Davide Bresolin on March, 2007
#

from ariadne import *

#set_hybrid_evolver_verbosity(6)
#set_integrator_verbosity(6)

max_time=5
g=-1

# Definition of the automaton
automaton=HybridAutomaton("Bouncing ball")

# Location fly:
mode_id=DiscreteState(0);
dyn=AffineVectorField(Matrix([[0, 1, 0],[ 0, 0, 0],[ 0, 0, 0]]), Vector([0,g,1]))
inv=RectangularSet([[0,2],[-2,1],[0,max_time]])
# inv=PolyhedralSet(Polyhedron(Matrix([[-1,0]]),Vector([0])))
l1=automaton.new_mode(mode_id, dyn, inv)

# Transition l1 -> l1
#act=Polyhedron(Matrix([[1,0],[ 0,1]]),Vector([0,0]))
event_id=DiscreteEvent(1)
act=RectangularSet([[-0.1,0.0],[-2,0],[0,max_time]])
res=AffineMap(Matrix([[1,0,0],[0,-0.5,0],[0,0,1]]), Vector([0,0,0]))
e12=automaton.new_transition(event_id,mode_id,mode_id,res,act)

print automaton

# Defintion of the underlying grid
grid=Grid(Vector([0.01,0.01,0.01]))

# Initial set 
init=Rectangle([[1.999,2],[0,0.001],[0,0.001]])
#init=Rectangle([[0,0.001],[0.999,1.0],[0,0.001]])
initial_set=HybridSet()
initial_set.new_location(mode_id,init)

print "initial_set.locations() =",initial_set.locations()
#print "initial_set[l0.id].size(),capacity()=",initial_set[l0.id()].size(),initial_set[l0.id()].capacity()

# Bounding set
bounding_box=Rectangle([[-0.1,2.1],[-2.1,1.1],[-0.1,max_time+0.1]]) 
bounding_set=HybridSet()
bounding_set.new_location(mode_id,bounding_box)
print "bounding_set.locations() =",bounding_set.locations()
#print "bounding_set[mode1_id].size(),capacity()=",bounding_set[m1.id()].size(),bounding_set[m1.id()].capacity()
#print "bounding_set[mode2_id].size(),capacity()=",bounding_set[m2.id()].size(),bounding_set[m2.id()].capacity()

# Definition of the Hybrid Evolver
par=EvolutionParameters()
par.set_maximum_step_size(0.5);
par.set_lock_to_grid_time(2.0);
par.set_grid(grid);
par.set_hybrid_bounding_domain(bounding_set);

apply=StandardApplicator()
integrator=AffineIntegrator();
#integrator=LohnerIntegrator(maximum_step_size,lock_to_grid_time,maximum_set_radius);
hybrid_evolver=SetBasedHybridEvolver(par,apply,integrator);

print "Computing reach set..."
#reach_set=hybrid_evolver.continuous_chainreach(automaton,initial_set,bounding_set)
reach_set=hybrid_evolver.chainreach(automaton,initial_set)

print "Done."

print "reach_set[l1.id].size() =",reach_set[mode_id].size()
print "reach_set[l1.id].capacity() =",reach_set[mode_id].capacity()

#print "Computing discrete transition..."
#trans_set=hybrid_evolver.discrete_step(automaton,reach_set)
#print "Done."

#print "trans_set[l1.id].size() =",trans_set[mode_id].size()
#print "trans_set[l1.id].capacity() =",trans_set[mode_id].capacity()

#print "Computing second continuous reach set..."
#reach_set2=hybrid_evolver.continuous_chainreach(automaton,trans_set,bounding_set)
#print "Done."

#print "reach_set2[l1.id].size() =",reach_set2[mode_id].size()
#print "reach_set2[l1.id].capacity() =",reach_set2[mode_id].capacity()

# Txt output
print "Creating txt output..."
txt=TextFile()
txt.open("bouncing-ball-3var.txt")

# Write the reached set
txt.write(reach_set[mode_id])
txt.close()

print "Done."

# Eps output
print "Creating eps output..."
eps=EpsPlot()
eps.open("bouncing-ball-3var.eps", bounding_box,2,0)

# Write the reached set
eps.write(reach_set[mode_id])
eps.close()

print "Done."





  	

File: examples/bouncing_ball_3var.cc
/***************************************************************************
 *            bouncing-ball-3var.cc
 *
 ****************************************************************************/


#include <iostream>
#include <fstream>
#include <string>

#include "ariadne.h"
#include "numeric/float64.h"
#include "geometry/set_interface.h"
#include "geometry/set_reference.h"
#include "geometry/hybrid_set.h"
#include "geometry/box.h"
#include "geometry/polyhedron.h"
#include "geometry/polyhedral_set.h"
#include "system/affine_map.h"
#include "system/affine_vector_field.h"
#include "system/hybrid_automaton.h"
#include "evaluation/evolution_parameters.h"
#include "evaluation/standard_applicator.h"
#include "evaluation/map_evolver.h"
#include "evaluation/vector_field_evolver.h"
#include "evaluation/lohner_integrator.h"
#include "evaluation/affine_integrator.h"
#include "evaluation/set_based_hybrid_evolver.h"
#include "output/epsstream.h"
#include "output/logging.h"

using namespace Ariadne;
using namespace Ariadne::Numeric;
using namespace Ariadne::LinearAlgebra;
using namespace Ariadne::Geometry;
using namespace Ariadne::System;
using namespace Ariadne::Evaluation;
using namespace Ariadne::Output;
using namespace Ariadne::Combinatoric;
using namespace std;

template<class R> int bouncing_ball_automaton();
  
int main() {
  return bouncing_ball_automaton<Float64>();
}

template<class R>
int bouncing_ball_automaton() 
{
  typedef Zonotope<R> BS;
  
  set_hybrid_evolver_verbosity(4);

  Box<R> domain("[0,15]x[-20,20]x[0,10]");
  cout << "domain=" << domain << endl;

  AffineVectorField<R> dynamic(Matrix<R>("[0,1,0;0,0,0;0,0,0]"),Vector<R>("[0,-9.8,1.0]"));
  cout << "dynamic=" << dynamic << endl;
  RectangularSet<R> invariant(domain);
  cout << "invariant=" << invariant << endl;

  RectangularSet<R> activation(Box<R>("[0.0,0.01]x[-20,20]x[0,10]"));
  cout << "activation=" << activation << endl;

  AffineMap<R> reset(Matrix<R>("[1,0,0;0,-1,0;0,0,1]"),Vector<R>("[0,0,0]"));
  cout << "reset=" << reset << endl;
  
  HybridAutomaton<R> automaton("Bouncing ball automaton");
  DiscreteState mode1_id(0);
  const DiscreteMode<R>& mode1=automaton.new_mode(mode1_id,dynamic,invariant);
  DiscreteEvent event_id(5);
  const DiscreteTransition<R>& transition=automaton.new_transition(event_id,mode1_id,mode1_id,reset,activation);
  
  cout << mode1  <<  "\n" << transition << endl;

  time_type maximum_step_size=0.125;
  time_type lock_to_grid_time=0.5;
  R maximum_set_radius=0.25;
  R grid_size=0.125;

  EvolutionParameters<R> parameters;
  parameters.set_maximum_step_size(0.125);
  parameters.set_lock_to_grid_time(0.5);
  parameters.set_maximum_basic_set_radius(0.25);
  parameters.set_grid_length(0.125);

  StandardApplicator<R> applicator;
	LohnerIntegrator<R> lohner;
  SetBasedHybridEvolver<BS> hybrid_evolver(parameters,applicator,lohner);
  
  Grid<R> grid(Vector<R>("[0.25,0.25,1.0]"));
  Box<R> bounding_box(domain.neighbourhood(1));
  FiniteGrid<R> finite_grid(grid,bounding_box);
  
  Box<R> initial_box("[10,10.1]x[0,0.1]x[0,0.1]");
  HybridGridMaskSet<R> initial_set;
  initial_set.new_location(mode1_id,finite_grid);
  initial_set[mode1_id].adjoin_over_approximation(initial_box);
  cout << "initial_set.discrete_locations()=" << initial_set.locations() << endl;
  cout << "initial_set[mode1_id].size()=" << initial_set[mode1_id].size() << " cells out of " << initial_set[mode1_id].capacity() << endl;
  
  cout << "Computing chainreachable set..." << endl;

  HybridGridMaskSet<R> chainreach=hybrid_evolver.upper_reach(automaton,initial_set,2.0);
  cout << "Reached (" << chainreach[mode1_id].size() << ") cells "
       << "out of (" << chainreach[mode1_id].capacity() << ") "
       << endl << endl;

  epsfstream eps; 

  eps.open("bouncing_ball-3var-xv.eps",bounding_box);
  eps << fill_colour(white) << bounding_box;
  eps << fill_colour(cyan) << activation;
  eps << line_style(true) << fill_colour(yellow) << chainreach[mode1_id];
  eps << fill_colour(blue) << initial_set[mode1_id];
  eps.close();
  
  eps.open("bouncing_ball-3var-tx.eps",bounding_box,PlanarProjectionMap(3,2,0));
  eps << fill_colour(white) << bounding_box;
  eps << fill_colour(cyan) << activation;
  eps << line_style(true) << fill_colour(yellow) << chainreach[mode1_id];
  eps << fill_colour(blue) << initial_set[mode1_id];
  eps.close();
  
  return 0;
}
